new-cdc/eventuate-local-java-cdc-connector-mysql-binlog/src/main/java/io/eventuate/local/mysql/binlog/MySqlBinaryLogClient.java

```diff
@@ -151,7 +139,33 @@ private void handleWriteRowsEvent(Event event, Optional<BinlogFileOffset> starti
       binlogEntryHandlers
               .stream()
               .filter(bh -> bh.isFor(database, table, defaultDatabase))
-              .forEach(bh -> bh.accept(eventData, binlogFilename, offset, startingBinlogFileOffset));
+              .forEach(new Consumer<BinlogEntryHandler>() {
+                private boolean couldReadDuplicateEntries = true;
+
+                @Override
+                public void accept(BinlogEntryHandler binlogEntryHandler) {
+                  try {
+                    MySqlBinlogEntryExtractor extractor = new MySqlBinlogEntryExtractor(dataSource,
+                            binlogEntryHandler.getSourceTableNameSupplier().getSourceTableName(),
+                            binlogEntryHandler.getEventuateSchema());
+
+                    BinlogEntry entry = extractor.extract(eventData, binlogFilename, offset);
+
+                    if (couldReadDuplicateEntries) {
+                      if (startingBinlogFileOffset.map(s -> s.isSameOrAfter(entry.getBinlogFileOffset())).orElse(false)) {
+                        return;
+                      } else {
+                        couldReadDuplicateEntries = false;
+                      }
+                    }
+
+                    binlogEntryHandler.publish(entry);
+
+                  } catch (IOException e) {
```

Comment:
Which method is throwing IOException?
Only that method should be wrapped in a catch block.
If its a method that you wrote then it should probably throw RuntimeException

Reply:
ok, I will fix